File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_cpp/URLHandler.cpp
#include <string>
#include <map>
#include <optional>
#include <sstream>
#include <vector>
#include <algorithm>

class URLHandler {
private:
    std::string url;

public:
    URLHandler(const std::string& url) : url(url) {}

    std::optional<std::string> getScheme() {
        size_t schemeEnd = url.find("://");
        if (schemeEnd != std::string::npos) {
            return url.substr(0, schemeEnd);
        }
        return std::nullopt;
    }

    std::optional<std::string> getHost() {
        size_t schemeEnd = url.find("://");
        if (schemeEnd != std::string::npos) {
            std::string urlWithoutScheme = url.substr(schemeEnd + 3);
            size_t hostEnd = urlWithoutScheme.find('/');
            if (hostEnd != std::string::npos) {
                return urlWithoutScheme.substr(0, hostEnd);
            }
            return urlWithoutScheme;
        }
        return std::nullopt;
    }

    std::optional<std::string> getPath() {
        size_t schemeEnd = url.find("://");
        if (schemeEnd != std::string::npos) {
            std::string urlWithoutScheme = url.substr(schemeEnd + 3);
            size_t hostEnd = urlWithoutScheme.find('/');
            if (hostEnd != std::string::npos) {
                return urlWithoutScheme.substr(hostEnd);
            }
        }
        return std::nullopt;
    }

    std::optional<std::map<std::string, std::string>> getQueryParams() {
        size_t queryStart = url.find('?');
        size_t fragmentStart = url.find('#');
        if (queryStart != std::string::npos) {
            std::string queryString;
            if (fragmentStart != std::string::npos) {
                queryString = url.substr(queryStart + 1, fragmentStart - (queryStart + 1));
            } else {
                queryString = url.substr(queryStart + 1);
            }

            if (!queryString.empty()) {
                std::map<std::string, std::string> params;
                std::istringstream ss(queryString);
                std::string pair;
                while (std::getline(ss, pair, '&')) {
                    size_t equalPos = pair.find('=');
                    if (equalPos != std::string::npos) {
                        std::string key = pair.substr(0, equalPos);
                        std::string value = pair.substr(equalPos + 1);
                        params[key] = value;
                    }
                }
                return params;
            }
        }
        return std::nullopt;
    }

    std::optional<std::string> getFragment() {
        size_t fragmentStart = url.find('#');
        if (fragmentStart != std::string::npos) {
            return url.substr(fragmentStart + 1);
        }
        return std::nullopt;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/cpp_to_py/code_output/URLHandler.py
class URLHandler:
    def __init__(self, url):
        self.url = url

    def get_scheme(self):
        scheme_end = self.url.find("://")
        if scheme_end != -1:
            return self.url[:scheme_end]
        return ""

    def get_host(self):
        scheme_end = self.url.find("://")
        if scheme_end != -1:
            url_without_scheme = self.url[scheme_end + 3:]
            host_end = url_without_scheme.find("/")
            if host_end != -1:
                return url_without_scheme[:host_end]
            return url_without_scheme
        return ""

    def get_path(self):
        scheme_end = self.url.find("://")
        if scheme_end != -1:
            url_without_scheme = self.url[scheme_end + 3:]
            host_end = url_without_scheme.find("/")
            if host_end != -1:
                return url_without_scheme[host_end:]
        return ""

    def get_query_params(self):
        params = {}
        query_start = self.url.find("?")
        fragment_start = self.url.find("#")
        if query_start != -1:
            if fragment_start != -1:
                query_string = self.url[query_start + 1:fragment_start]
            else:
                query_string = self.url[query_start + 1:]
            if query_string:
                pos = query_string.find("&")
                while pos != -1:
                    token = query_string[:pos]
                    equal_pos = token.find("=")
                    if equal_pos != -1:
                        key = token[:equal_pos]
                        value = token[equal_pos + 1:]
                        params[key] = value
                    query_string = query_string[pos + 1:]
                    pos = query_string.find("&")
                equal_pos = query_string.find("=")
                if equal_pos != -1:
                    key = query_string[:equal_pos]
                    value = query_string[equal_pos + 1:]
                    params[key] = value
        return params

    def get_fragment(self):
        fragment_start = self.url.find("#")
        if fragment_start != -1:
            return self.url[fragment_start + 1:]
        return ""